File: IN3130/Oblig2/src/test/java/assignment/AStarTest.java
package assignment;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class AStarTest {

    private AStar astar;

    @BeforeEach
    public void setUp() {
        astar = new AStar();
    }

    @Test
    public void testCase1() {
        int[][] input = {
                {1, 2, 0},
                {4, 6, 3},
                {7, 5, 8}
        };
        String solution = astar.solve(input);
        String suggestedSolution = "DLDR";
        assertEquals(suggestedSolution.length(), solution.length(), "Solution is not optimal");
        assertTrue(isSolutionValid(input, solution), "The puzzle is not solved after applying the solution moves.");
    }

    @Test
    public void testCase2() {
        int[][] input = {
                {1, 0, 5},
                {4, 3, 2},
                {7, 8, 6}
        };
        String solution = astar.solve(input);
        String suggestedSolution = "DRULDRD";
        assertEquals(suggestedSolution.length(), solution.length(), "Solution is not optimal");
        assertTrue(isSolutionValid(input, solution), "The puzzle is not solved after applying the solution moves.");
    }

    @Test
    public void testCase3() {
        int[][] input = {
                {0, 1, 7},
                {5, 2, 6},
                {4, 8, 3}
        };
        String solution = astar.solve(input);
        String suggestedSolution = "RRDLURDDLUURDLLDRR";
        assertEquals(suggestedSolution.length(), solution.length(), "Solution is not optimal");
        assertTrue(isSolutionValid(input, solution), "The puzzle is not solved after applying the solution moves.");
    }

    @Test
    public void testCase4() {
        int[][] input = {
                {0, 1, 3, 4},
                {5, 2, 6, 7},
                {9, 10, 11, 8},
                {13, 14, 15, 12}
        };
        String solution = astar.solve(input);
        String suggestedSolution = "RDRRDD";
        assertEquals(suggestedSolution.length(), solution.length(), "Solution is not optimal");
        assertTrue(isSolutionValid(input, solution), "The puzzle is not solved after applying the solution moves.");
    }

    private boolean isSolutionValid(int[][] board, String moves) {
        int[][] clonedBoard = cloneBoard(board);
        applyMoves(clonedBoard, moves);
        return isSolved(clonedBoard);
    }

    private int[][] cloneBoard(int[][] board) {
        int n = board.length;
        int[][] cloned = new int[n][n];
        for (int i = 0; i < n; i++) {
            System.arraycopy(board[i], 0, cloned[i], 0, n);
        }
        return cloned;
    }

    private void applyMoves(int[][] board, String moves) {
        int n = board.length;
        int x = 0, y = 0;
        search:
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 0) {
                    x = i;
                    y = j;
                    break search;
                }
            }
        }

        for (char move : moves.toCharArray()) {
            switch (move) {
                case 'L' -> {
                    if (y > 0) {
                        swap(board, x, y, x, y - 1);
                        y--;
                    }
                }
                case 'R' -> {
                    if (y < n - 1) {
                        swap(board, x, y, x, y + 1);
                        y++;
                    }
                }
                case 'U' -> {
                    if (x > 0) {
                        swap(board, x, y, x - 1, y);
                        x--;
                    }
                }
                case 'D' -> {
                    if (x < n - 1) {
                        swap(board, x, y, x + 1, y);
                        x++;
                    }
                }
            }
        }
    }

    private boolean isSolved(int[][] board) {
        int n = board.length;
        int expectedValue = 1;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == n - 1 && j == n - 1) {
                    // The last cell should be 0
                    if (board[i][j] != 0) return false;
                } else {
                    if (board[i][j] != expectedValue) return false;
                    expectedValue++;
                }
            }
        }

        return true;
    }

    private void swap(int[][] board, int x1, int y1, int x2, int y2) {
        int temp = board[x1][y1];
        board[x1][y1] = board[x2][y2];
        board[x2][y2] = temp;
    }
}

File: IN3130/Oblig2/src/python/a_star_test.py
import pytest
import numpy as np

from AStar import solve


def is_solution_valid(board, moves):
    cloned_board = np.copy(board)
    apply_moves(cloned_board, moves)
    return is_solved(cloned_board)


def apply_moves(board, moves):
    n = board.shape[0]
    x, y = np.argwhere(board == 0)[0]
    for move in moves:
        if move == 'L' and y > 0:
            board[x, y], board[x, y - 1] = board[x, y - 1], board[x, y]
            y -= 1
        elif move == 'R' and y < n - 1:
            board[x, y], board[x, y + 1] = board[x, y + 1], board[x, y]
            y += 1
        elif move == 'U' and x > 0:
            board[x, y], board[x - 1, y] = board[x - 1, y], board[x, y]
            x -= 1
        elif move == 'D' and x < n - 1:
            board[x, y], board[x + 1, y] = board[x + 1, y], board[x, y]
            x += 1


def is_solved(board):
    n = board.shape[0]
    expected_value = 1
    for i in range(n):
        for j in range(n):
            if i == n - 1 and j == n - 1:
                if board[i, j] != 0:
                    return False
            else:
                if board[i, j] != expected_value:
                    return False
                expected_value += 1
    return True


class TestAStar:
    def test_case1(self):
        input_board = np.array([
            [1, 2, 0],
            [4, 6, 3],
            [7, 5, 8]
        ])
        solution = solve(input_board)
        suggested_solution = "DLDR"
        assert len(suggested_solution) == len(solution), "Solution is not optimal"
        assert is_solution_valid(input_board, solution), "The puzzle is not solved after applying the solution moves."

    def test_case2(self):
        input_board = np.array([
            [1, 0, 5],
            [4, 3, 2],
            [7, 8, 6]
        ])
        solution = solve(input_board)
        suggested_solution = "DRULDRD"
        assert len(suggested_solution) == len(solution), "Solution is not optimal"
        assert is_solution_valid(input_board, solution), "The puzzle is not solved after applying the solution moves."

    def test_case3(self):
        input_board = np.array([
            [0, 1, 7],
            [5, 2, 6],
            [4, 8, 3]]
        )
        solution = solve(input_board)
        suggested_solution = "RRDLURDDLUURDLLDRR"
        assert len(suggested_solution) == len(solution), "Solution is not optimal"
        assert is_solution_valid(input_board, solution), "The puzzle is not solved after applying the solution moves."

    def test_case4(self):
        input_board = np.array([
            [0, 1, 3, 4],
            [5, 2, 6, 7],
            [9, 10, 11, 8],
            [13, 14, 15, 12]
        ])
        solution = solve(input_board)
        suggested_solution = "RDRRDD"
        assert len(suggested_solution) == len(solution), "Solution is not optimal"
        assert is_solution_valid(input_board, solution), "The puzzle is not solved after applying the solution moves."
